// hdmi-cec/dist/models/logical-address.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var LogicalAddress;
(function (LogicalAddress) {
    LogicalAddress[LogicalAddress["UNKNOWN"] = -1] = "UNKNOWN";
    LogicalAddress[LogicalAddress["TV"] = 0] = "TV";
    LogicalAddress[LogicalAddress["RECORDINGDEVICE1"] = 1] = "RECORDINGDEVICE1";
    LogicalAddress[LogicalAddress["RECORDINGDEVICE2"] = 2] = "RECORDINGDEVICE2";
    LogicalAddress[LogicalAddress["TUNER1"] = 3] = "TUNER1";
    LogicalAddress[LogicalAddress["PLAYBACKDEVICE1"] = 4] = "PLAYBACKDEVICE1";
    LogicalAddress[LogicalAddress["AUDIOSYSTEM"] = 5] = "AUDIOSYSTEM";
    LogicalAddress[LogicalAddress["TUNER2"] = 6] = "TUNER2";
    LogicalAddress[LogicalAddress["TUNER3"] = 7] = "TUNER3";
    LogicalAddress[LogicalAddress["PLAYBACKDEVICE2"] = 8] = "PLAYBACKDEVICE2";
    LogicalAddress[LogicalAddress["RECORDINGDEVICE3"] = 9] = "RECORDINGDEVICE3";
    LogicalAddress[LogicalAddress["TUNER4"] = 10] = "TUNER4";
    LogicalAddress[LogicalAddress["PLAYBACKDEVICE3"] = 11] = "PLAYBACKDEVICE3";
    LogicalAddress[LogicalAddress["RESERVED1"] = 12] = "RESERVED1";
    LogicalAddress[LogicalAddress["RESERVED2"] = 13] = "RESERVED2";
    LogicalAddress[LogicalAddress["FREEUSE"] = 14] = "FREEUSE";
    LogicalAddress[LogicalAddress["UNREGISTERED"] = 15] = "UNREGISTERED";
    LogicalAddress[LogicalAddress["BROADCAST"] = 15] = "BROADCAST"; // for target
})(LogicalAddress = exports.LogicalAddress || (exports.LogicalAddress = {}));
