// hdmi-cec/dist/models/operation-code.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var OperationCode;
(function (OperationCode) {
    OperationCode[OperationCode["ACTIVE_SOURCE"] = 130] = "ACTIVE_SOURCE";
    OperationCode[OperationCode["IMAGE_VIEW_ON"] = 4] = "IMAGE_VIEW_ON";
    OperationCode[OperationCode["TEXT_VIEW_ON"] = 13] = "TEXT_VIEW_ON";
    OperationCode[OperationCode["INACTIVE_SOURCE"] = 157] = "INACTIVE_SOURCE";
    OperationCode[OperationCode["REQUEST_ACTIVE_SOURCE"] = 133] = "REQUEST_ACTIVE_SOURCE";
    OperationCode[OperationCode["ROUTING_CHANGE"] = 128] = "ROUTING_CHANGE";
    OperationCode[OperationCode["ROUTING_INFORMATION"] = 129] = "ROUTING_INFORMATION";
    OperationCode[OperationCode["SET_STREAM_PATH"] = 134] = "SET_STREAM_PATH";
    OperationCode[OperationCode["STANDBY"] = 54] = "STANDBY";
    OperationCode[OperationCode["RECORD_OFF"] = 11] = "RECORD_OFF";
    OperationCode[OperationCode["RECORD_ON"] = 9] = "RECORD_ON";
    OperationCode[OperationCode["RECORD_STATUS"] = 10] = "RECORD_STATUS";
    OperationCode[OperationCode["RECORD_TV_SCREEN"] = 15] = "RECORD_TV_SCREEN";
    OperationCode[OperationCode["CLEAR_ANALOGUE_TIMER"] = 51] = "CLEAR_ANALOGUE_TIMER";
    OperationCode[OperationCode["CLEAR_DIGITAL_TIMER"] = 153] = "CLEAR_DIGITAL_TIMER";
    OperationCode[OperationCode["CLEAR_EXTERNAL_TIMER"] = 161] = "CLEAR_EXTERNAL_TIMER";
    OperationCode[OperationCode["SET_ANALOGUE_TIMER"] = 52] = "SET_ANALOGUE_TIMER";
    OperationCode[OperationCode["SET_DIGITAL_TIMER"] = 151] = "SET_DIGITAL_TIMER";
    OperationCode[OperationCode["SET_EXTERNAL_TIMER"] = 162] = "SET_EXTERNAL_TIMER";
    OperationCode[OperationCode["SET_TIMER_PROGRAM_TITLE"] = 103] = "SET_TIMER_PROGRAM_TITLE";
    OperationCode[OperationCode["TIMER_CLEARED_STATUS"] = 67] = "TIMER_CLEARED_STATUS";
    OperationCode[OperationCode["TIMER_STATUS"] = 53] = "TIMER_STATUS";
    OperationCode[OperationCode["CEC_VERSION"] = 158] = "CEC_VERSION";
    OperationCode[OperationCode["GET_CEC_VERSION"] = 159] = "GET_CEC_VERSION";
    OperationCode[OperationCode["GIVE_PHYSICAL_ADDRESS"] = 131] = "GIVE_PHYSICAL_ADDRESS";
    OperationCode[OperationCode["GET_MENU_LANGUAGE"] = 145] = "GET_MENU_LANGUAGE";
    OperationCode[OperationCode["REPORT_PHYSICAL_ADDRESS"] = 132] = "REPORT_PHYSICAL_ADDRESS";
    OperationCode[OperationCode["SET_MENU_LANGUAGE"] = 50] = "SET_MENU_LANGUAGE";
    OperationCode[OperationCode["DECK_CONTROL"] = 66] = "DECK_CONTROL";
    OperationCode[OperationCode["DECK_STATUS"] = 27] = "DECK_STATUS";
    OperationCode[OperationCode["GIVE_DECK_STATUS"] = 26] = "GIVE_DECK_STATUS";
    OperationCode[OperationCode["PLAY"] = 65] = "PLAY";
    OperationCode[OperationCode["GIVE_TUNER_DEVICE_STATUS"] = 8] = "GIVE_TUNER_DEVICE_STATUS";
    OperationCode[OperationCode["SELECT_ANALOGUE_SERVICE"] = 146] = "SELECT_ANALOGUE_SERVICE";
    OperationCode[OperationCode["SELECT_DIGITAL_SERVICE"] = 147] = "SELECT_DIGITAL_SERVICE";
    OperationCode[OperationCode["TUNER_DEVICE_STATUS"] = 7] = "TUNER_DEVICE_STATUS";
    OperationCode[OperationCode["TUNER_STEP_DECREMENT"] = 6] = "TUNER_STEP_DECREMENT";
    OperationCode[OperationCode["TUNER_STEP_INCREMENT"] = 5] = "TUNER_STEP_INCREMENT";
    OperationCode[OperationCode["DEVICE_VENDOR_ID"] = 135] = "DEVICE_VENDOR_ID";
    OperationCode[OperationCode["GIVE_DEVICE_VENDOR_ID"] = 140] = "GIVE_DEVICE_VENDOR_ID";
    OperationCode[OperationCode["VENDOR_COMMAND"] = 137] = "VENDOR_COMMAND";
    OperationCode[OperationCode["VENDOR_COMMAND_WITH_ID"] = 160] = "VENDOR_COMMAND_WITH_ID";
    OperationCode[OperationCode["VENDOR_REMOTE_BUTTON_DOWN"] = 138] = "VENDOR_REMOTE_BUTTON_DOWN";
    OperationCode[OperationCode["VENDOR_REMOTE_BUTTON_UP"] = 139] = "VENDOR_REMOTE_BUTTON_UP";
    OperationCode[OperationCode["SET_OSD_STRING"] = 100] = "SET_OSD_STRING";
    OperationCode[OperationCode["GIVE_OSD_NAME"] = 70] = "GIVE_OSD_NAME";
    OperationCode[OperationCode["SET_OSD_NAME"] = 71] = "SET_OSD_NAME";
    OperationCode[OperationCode["MENU_REQUEST"] = 141] = "MENU_REQUEST";
    OperationCode[OperationCode["MENU_STATUS"] = 142] = "MENU_STATUS";
    OperationCode[OperationCode["USER_CONTROL_PRESSED"] = 68] = "USER_CONTROL_PRESSED";
    OperationCode[OperationCode["USER_CONTROL_RELEASE"] = 69] = "USER_CONTROL_RELEASE";
    OperationCode[OperationCode["GIVE_DEVICE_POWER_STATUS"] = 143] = "GIVE_DEVICE_POWER_STATUS";
    OperationCode[OperationCode["REPORT_POWER_STATUS"] = 144] = "REPORT_POWER_STATUS";
    OperationCode[OperationCode["FEATURE_ABORT"] = 0] = "FEATURE_ABORT";
    OperationCode[OperationCode["ABORT"] = 255] = "ABORT";
    OperationCode[OperationCode["GIVE_AUDIO_STATUS"] = 113] = "GIVE_AUDIO_STATUS";
    OperationCode[OperationCode["GIVE_SYSTEM_AUDIO_MODE_STATUS"] = 125] = "GIVE_SYSTEM_AUDIO_MODE_STATUS";
    OperationCode[OperationCode["REPORT_AUDIO_STATUS"] = 122] = "REPORT_AUDIO_STATUS";
    OperationCode[OperationCode["SET_SYSTEM_AUDIO_MODE"] = 114] = "SET_SYSTEM_AUDIO_MODE";
    OperationCode[OperationCode["SYSTEM_AUDIO_MODE_REQUEST"] = 112] = "SYSTEM_AUDIO_MODE_REQUEST";
    OperationCode[OperationCode["SYSTEM_AUDIO_MODE_STATUS"] = 126] = "SYSTEM_AUDIO_MODE_STATUS";
    OperationCode[OperationCode["SET_AUDIO_RATE"] = 154] = "SET_AUDIO_RATE";
    // CEC 1.4
    OperationCode[OperationCode["START_ARC"] = 192] = "START_ARC";
    OperationCode[OperationCode["REPORT_ARC_STARTED"] = 193] = "REPORT_ARC_STARTED";
    OperationCode[OperationCode["REPORT_ARC_ENDED"] = 194] = "REPORT_ARC_ENDED";
    OperationCode[OperationCode["REQUEST_ARC_START"] = 195] = "REQUEST_ARC_START";
    OperationCode[OperationCode["REQUEST_ARC_END"] = 196] = "REQUEST_ARC_END";
    OperationCode[OperationCode["END_ARC"] = 197] = "END_ARC";
    OperationCode[OperationCode["CDC"] = 248] = "CDC";
    // when this opcode is set no opcode will be sent to the device. this is one of the reserved numbers
    OperationCode[OperationCode["NONE"] = 253] = "NONE";
})(OperationCode = exports.OperationCode || (exports.OperationCode = {}));
