// hdmi-cec/dist/models/user-control-buttons.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var UserControlButton;
(function (UserControlButton) {
    UserControlButton[UserControlButton["SELECT"] = 0] = "SELECT";
    UserControlButton[UserControlButton["UP"] = 1] = "UP";
    UserControlButton[UserControlButton["DOWN"] = 2] = "DOWN";
    UserControlButton[UserControlButton["LEFT"] = 3] = "LEFT";
    UserControlButton[UserControlButton["RIGHT"] = 4] = "RIGHT";
    UserControlButton[UserControlButton["RIGHT_UP"] = 5] = "RIGHT_UP";
    UserControlButton[UserControlButton["RIGHT_DOWN"] = 6] = "RIGHT_DOWN";
    UserControlButton[UserControlButton["LEFT_UP"] = 7] = "LEFT_UP";
    UserControlButton[UserControlButton["LEFT_DOWN"] = 8] = "LEFT_DOWN";
    UserControlButton[UserControlButton["ROOT_MENU"] = 9] = "ROOT_MENU";
    UserControlButton[UserControlButton["SETUP_MENU"] = 10] = "SETUP_MENU";
    UserControlButton[UserControlButton["CONTENTS_MENU"] = 11] = "CONTENTS_MENU";
    UserControlButton[UserControlButton["FAVORITE_MENU"] = 12] = "FAVORITE_MENU";
    UserControlButton[UserControlButton["EXIT"] = 13] = "EXIT";
    // reserved = 0x0E, 0x0F
    UserControlButton[UserControlButton["TOP_MENU"] = 16] = "TOP_MENU";
    UserControlButton[UserControlButton["DVD_MENU"] = 17] = "DVD_MENU";
    // reserved = 0x12 ... 0x1C
    UserControlButton[UserControlButton["NUMBER_ENTRY_MODE"] = 29] = "NUMBER_ENTRY_MODE";
    UserControlButton[UserControlButton["NUMBER11"] = 30] = "NUMBER11";
    UserControlButton[UserControlButton["NUMBER12"] = 31] = "NUMBER12";
    UserControlButton[UserControlButton["NUMBER0"] = 32] = "NUMBER0";
    UserControlButton[UserControlButton["NUMBER1"] = 33] = "NUMBER1";
    UserControlButton[UserControlButton["NUMBER2"] = 34] = "NUMBER2";
    UserControlButton[UserControlButton["NUMBER3"] = 35] = "NUMBER3";
    UserControlButton[UserControlButton["NUMBER4"] = 36] = "NUMBER4";
    UserControlButton[UserControlButton["NUMBER5"] = 37] = "NUMBER5";
    UserControlButton[UserControlButton["NUMBER6"] = 38] = "NUMBER6";
    UserControlButton[UserControlButton["NUMBER7"] = 39] = "NUMBER7";
    UserControlButton[UserControlButton["NUMBER8"] = 40] = "NUMBER8";
    UserControlButton[UserControlButton["NUMBER9"] = 41] = "NUMBER9";
    UserControlButton[UserControlButton["DOT"] = 42] = "DOT";
    UserControlButton[UserControlButton["ENTER"] = 43] = "ENTER";
    UserControlButton[UserControlButton["CLEAR"] = 44] = "CLEAR";
    UserControlButton[UserControlButton["NEXT_FAVORITE"] = 47] = "NEXT_FAVORITE";
    UserControlButton[UserControlButton["CHANNEL_UP"] = 48] = "CHANNEL_UP";
    UserControlButton[UserControlButton["CHANNEL_DOWN"] = 49] = "CHANNEL_DOWN";
    UserControlButton[UserControlButton["PREVIOUS_CHANNEL"] = 50] = "PREVIOUS_CHANNEL";
    UserControlButton[UserControlButton["SOUND_SELECT"] = 51] = "SOUND_SELECT";
    UserControlButton[UserControlButton["INPUT_SELECT"] = 52] = "INPUT_SELECT";
    UserControlButton[UserControlButton["DISPLAY_INFORMATION"] = 53] = "DISPLAY_INFORMATION";
    UserControlButton[UserControlButton["HELP"] = 54] = "HELP";
    UserControlButton[UserControlButton["PAGE_UP"] = 55] = "PAGE_UP";
    UserControlButton[UserControlButton["PAGE_DOWN"] = 56] = "PAGE_DOWN";
    // reserved = 0x39 ... 0x3F
    UserControlButton[UserControlButton["POWER"] = 64] = "POWER";
    UserControlButton[UserControlButton["VOLUME_UP"] = 65] = "VOLUME_UP";
    UserControlButton[UserControlButton["VOLUME_DOWN"] = 66] = "VOLUME_DOWN";
    UserControlButton[UserControlButton["MUTE"] = 67] = "MUTE";
    UserControlButton[UserControlButton["PLAY"] = 68] = "PLAY";
    UserControlButton[UserControlButton["STOP"] = 69] = "STOP";
    UserControlButton[UserControlButton["PAUSE"] = 70] = "PAUSE";
    UserControlButton[UserControlButton["RECORD"] = 71] = "RECORD";
    UserControlButton[UserControlButton["REWIND"] = 72] = "REWIND";
    UserControlButton[UserControlButton["FAST_FORWARD"] = 73] = "FAST_FORWARD";
    UserControlButton[UserControlButton["EJECT"] = 74] = "EJECT";
    UserControlButton[UserControlButton["FORWARD"] = 75] = "FORWARD";
    UserControlButton[UserControlButton["BACKWARD"] = 76] = "BACKWARD";
    UserControlButton[UserControlButton["STOP_RECORD"] = 77] = "STOP_RECORD";
    UserControlButton[UserControlButton["PAUSE_RECORD"] = 78] = "PAUSE_RECORD";
    // reserved = 0x4F
    UserControlButton[UserControlButton["ANGLE"] = 80] = "ANGLE";
    UserControlButton[UserControlButton["SUB_PICTURE"] = 81] = "SUB_PICTURE";
    UserControlButton[UserControlButton["VIDEO_ON_DEMAND"] = 82] = "VIDEO_ON_DEMAND";
    UserControlButton[UserControlButton["ELECTRONIC_PROGRAM_GUIDE"] = 83] = "ELECTRONIC_PROGRAM_GUIDE";
    UserControlButton[UserControlButton["TIMER_PROGRAMMING"] = 84] = "TIMER_PROGRAMMING";
    UserControlButton[UserControlButton["INITIAL_CONFIGURATION"] = 85] = "INITIAL_CONFIGURATION";
    UserControlButton[UserControlButton["SELECT_BROADCAST_TYPE"] = 86] = "SELECT_BROADCAST_TYPE";
    UserControlButton[UserControlButton["SELECT_SOUND_PRESENTATION"] = 87] = "SELECT_SOUND_PRESENTATION";
    // reserved = 0x58 ... 0x5F
    UserControlButton[UserControlButton["PLAY_FUNCTION"] = 96] = "PLAY_FUNCTION";
    UserControlButton[UserControlButton["PAUSE_PLAY_FUNCTION"] = 97] = "PAUSE_PLAY_FUNCTION";
    UserControlButton[UserControlButton["RECORD_FUNCTION"] = 98] = "RECORD_FUNCTION";
    UserControlButton[UserControlButton["PAUSE_RECORD_FUNCTION"] = 99] = "PAUSE_RECORD_FUNCTION";
    UserControlButton[UserControlButton["STOP_FUNCTION"] = 100] = "STOP_FUNCTION";
    UserControlButton[UserControlButton["MUTE_FUNCTION"] = 101] = "MUTE_FUNCTION";
    UserControlButton[UserControlButton["RESTORE_VOLUME_FUNCTION"] = 102] = "RESTORE_VOLUME_FUNCTION";
    UserControlButton[UserControlButton["TUNE_FUNCTION"] = 103] = "TUNE_FUNCTION";
    UserControlButton[UserControlButton["SELECT_MEDIA_FUNCTION"] = 104] = "SELECT_MEDIA_FUNCTION";
    UserControlButton[UserControlButton["SELECT_AV_INPUT_FUNCTION"] = 105] = "SELECT_AV_INPUT_FUNCTION";
    UserControlButton[UserControlButton["SELECT_AUDIO_INPUT_FUNCTION"] = 106] = "SELECT_AUDIO_INPUT_FUNCTION";
    UserControlButton[UserControlButton["POWER_TOGGLE_FUNCTION"] = 107] = "POWER_TOGGLE_FUNCTION";
    UserControlButton[UserControlButton["POWER_OFF_FUNCTION"] = 108] = "POWER_OFF_FUNCTION";
    UserControlButton[UserControlButton["POWER_ON_FUNCTION"] = 109] = "POWER_ON_FUNCTION";
    // reserved = 0x6E ... 0x70
    UserControlButton[UserControlButton["F1_BLUE"] = 113] = "F1_BLUE";
    UserControlButton[UserControlButton["F2_RED"] = 114] = "F2_RED";
    UserControlButton[UserControlButton["F3_GREEN"] = 115] = "F3_GREEN";
    UserControlButton[UserControlButton["F4_YELLOW"] = 116] = "F4_YELLOW";
    UserControlButton[UserControlButton["F5"] = 117] = "F5";
    UserControlButton[UserControlButton["DATA"] = 118] = "DATA";
    // reserved = 0x77 ... 0xFF
    UserControlButton[UserControlButton["AN_RETURN"] = 145] = "AN_RETURN";
    UserControlButton[UserControlButton["AN_CHANNELS_LIST"] = 150] = "AN_CHANNELS_LIST";
    UserControlButton[UserControlButton["MAX"] = 150] = "MAX";
    UserControlButton[UserControlButton["UNKNOWN"] = 255] = "UNKNOWN";
})(UserControlButton = exports.UserControlButton || (exports.UserControlButton = {}));
